src/bit-systems/audio-emitter-system.ts

```diff
@@ -0,0 +1,113 @@
+import { addComponent, addEntity, defineQuery, removeComponent } from "bitecs";
+import { PositionalAudio, Audio as StereoAudio, AudioListener as ThreeAudioListener } from "three";
+import { HubsWorld } from "../app";
+import { AudioEmitter, AudioSettingsChanged } from "../bit-components";
+import { AudioType, SourceType } from "../components/audio-params";
+import { AudioSystem } from "../systems/audio-system";
+import { applySettings, getCurrentAudioSettings, updateAudioSettings } from "../update-audio-settings";
+import { addObject3DComponent, swapObject3DComponent } from "../utils/jsx-entity";
+
+export type AudioObject3D = StereoAudio | PositionalAudio;
+type AudioConstructor<T> = new (listener: ThreeAudioListener) => T;
+
+export const Emitter2Audio = (AudioEmitter as any).audios as Map<number, number>;
+export const Emitter2Params = (AudioEmitter as any).params as Map<number, number>;
+
+export const EMITTER_FLAGS = {
+  MUTED: 1 << 0,
+  PAUSED: 1 << 1,
+  CLIPPED: 1 << 2
+};
+
+export function isPositionalAudio(node: AudioObject3D): node is PositionalAudio {
+  return (node as any).panner !== undefined;
+}
+
+export function cleanupAudio(audio: AudioObject3D) {
+  const eid = audio.eid!;
+  audio.disconnect();
+  const audioSystem = APP.scene?.systems["hubs-systems"].audioSystem;
+  APP.audios.delete(eid);
+  APP.supplementaryAttenuation.delete(eid);
+  APP.audioOverrides.delete(eid);
+  audioSystem.removeAudio({ node: audio });
+}
+
+function swapAudioType<T extends AudioObject3D>(
+  world: HubsWorld,
+  audioSystem: AudioSystem,
+  eid: number,
+  NewType: AudioConstructor<T>
+) {
+  const audio = world.eid2obj.get(eid)! as AudioObject3D;
+  audio.disconnect();
+  APP.gainMultipliers.delete(eid);
+  APP.sourceType.set(eid, SourceType.MEDIA_VIDEO);
+  APP.supplementaryAttenuation.delete(eid);
+  APP.audioOverrides.delete(eid);
+  APP.audios.delete(eid);
+  audioSystem.removeAudio({ node: audio });
+
+  const newAudio = new NewType(APP.audioListener);
+  newAudio.setNodeSource(audio.source!);
+  audioSystem.addAudio({ sourceType: SourceType.MEDIA_VIDEO, node: newAudio });
+  APP.audios.set(eid, newAudio);
+
+  audio.parent!.add(newAudio);
+  audio.removeFromParent();
+
+  swapObject3DComponent(world, eid, newAudio);
```

Comment:
I know this is not new, but this the main complexity we incure by using three.js audio / positional audio nodes instead of using the webaudio api directly. I believe this is the only place in the app where we swap an entity id from one object3D to another. It would be good to revisit this (not in this PR).

Reply:
Yep, I agree. The goal is to address when I'm do migrating all the services. In but-ecs all the Audio logic should be living in the `audio-emitter-system.ts` so it should be easier to remove that indirection layer when we have migrated all the systems/components.